File: apps/mobile/src/db/models/Goal.ts
import { Model } from '@nozbe/watermelondb';
import { field, readonly, date, text } from '@nozbe/watermelondb/decorators';

export class Goal extends Model {
  static table = 'goals';

  @text('user_id')        userId!: string;
  @text('name')           name!: string;
  @field('target_amount') targetAmount!: number;
  @field('current_amount') currentAmount!: number;
  @text('currency')       currency!: string;
  @field('deadline')      deadline!: number;
  @text('type')           type!: 'savings' | 'debt' | 'emergency' | 'investment' | 'custom';
  @field('is_completed')  isCompleted!: boolean;
  @field('is_synced')     isSynced!: boolean;
  @text('server_id')      serverId!: string;

  @readonly @date('created_at') createdAt!: Date;
  @readonly @date('updated_at') updatedAt!: Date;

  get progressPercent(): number {
    if (this.targetAmount === 0) return 0;
    return Math.min((this.currentAmount / this.targetAmount) * 100, 100);
  }
}

File: apps/api/app/models/goal.py
import uuid
import enum
from datetime import datetime
from sqlalchemy import String, Numeric, Boolean, ForeignKey, DateTime, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func
from app.database import Base


class GoalType(str, enum.Enum):
    savings    = "savings"
    debt       = "debt"
    emergency  = "emergency"
    investment = "investment"
    custom     = "custom"


class Goal(Base):
    __tablename__ = "goals"

    id:             Mapped[str]      = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id:        Mapped[str]      = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    name:           Mapped[str]      = mapped_column(String(255), nullable=False)
    target_amount:  Mapped[float]    = mapped_column(Numeric(14, 2), nullable=False)
    current_amount: Mapped[float]    = mapped_column(Numeric(14, 2), default=0)
    currency:       Mapped[str]      = mapped_column(String(3), nullable=False)
    deadline:       Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    type:           Mapped[GoalType] = mapped_column(SAEnum(GoalType), default=GoalType.savings)
    is_completed:   Mapped[bool]     = mapped_column(Boolean, default=False)
    created_at:     Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at:     Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    user = relationship("User", back_populates="goals")
